aspnetcore-authentication-jwtbearer/test/AspNetCore.Authentication.JwtBearer.Tests/DPoP/DPoPIntegrationTests.cs

```diff
@@ -0,0 +1,396 @@
+
+        // Verify that a custom nonce was issued in the response
+        result.Headers.TryGetValues("DPoP-Nonce", out var nonceValues).ShouldBeTrue();
+        var serverNonce = nonceValues?.FirstOrDefault();
+        serverNonce.ShouldNotBeNull();
+        serverNonce.ShouldStartWith("custom-nonce-");
+
+        // Make a new request with the custom server issued nonce
+        var proofWithNonce = await CreateProofToken(token, Jwk, HttpMethod.Get, new Uri("http://localhost/"),
+            nonce: DPoPNonce.Parse(serverNonce));
+        Api.HttpClient.DefaultRequestHeaders.Remove(OidcConstants.HttpHeaders.DPoP);
+        Api.HttpClient.DefaultRequestHeaders.Add(OidcConstants.HttpHeaders.DPoP, proofWithNonce);
+        var secondResult = await Api.HttpClient.GetAsync("/");
+        secondResult.StatusCode.ShouldBe(HttpStatusCode.OK);
+    }
+
+    [Fact]
+    public async Task replay_cache_can_be_customized_by_replacing_keyed_hybrid_cache()
+    {
+        var cache = new TestHybridCache();
+        Api.OnConfigureServices += services =>
+        {
+            services.AddKeyedSingleton<HybridCache>(ServiceProviderKeys.ProofTokenReplayHybridCache, cache);
+        };
+        ApiOptions = opt => opt.EnableReplayDetection = true;
+        await Initialize();
+
+        var token = await LoginAndGetToken();
+        Api.HttpClient.SetToken("DPoP", token.AccessToken);
+
+        var proof = await CreateProofToken(token, Jwk, HttpMethod.Get, new Uri("http://localhost/"));
+        Api.HttpClient.DefaultRequestHeaders.Add(OidcConstants.HttpHeaders.DPoP, proof);
+
+        var result = await Api.HttpClient.GetAsync("/");
+
+        result.StatusCode.ShouldBe(HttpStatusCode.OK);
+
+        // Verify that the custom cache was used for replay detection
+        cache.SetAsyncCalls.Count.ShouldBeGreaterThan(0);
+        cache.GetOrCreateAsyncCalls.Count.ShouldBeGreaterThan(0);
+    }
+
+    [Fact]
+    public async Task replay_cache_can_be_customized_by_replacing_keyed_distributed_cache()
+    {
+        var cache = new TestDistributedCache();
+        Api.OnConfigureServices += services =>
+        {
+            services.AddKeyedSingleton<IDistributedCache>(ServiceProviderKeys.ProofTokenReplayDistributedCache, cache);
+        };
+        ApiOptions = opt => opt.EnableReplayDetection = true;
+        await Initialize();
+
+        var token = await LoginAndGetToken();
+        Api.HttpClient.SetToken("DPoP", token.AccessToken);
+
+        var proof = await CreateProofToken(token, Jwk, HttpMethod.Get, new Uri("http://localhost/"));
+        Api.HttpClient.DefaultRequestHeaders.Add(OidcConstants.HttpHeaders.DPoP, proof);
+
+        var result = await Api.HttpClient.GetAsync("/");
+
+        result.StatusCode.ShouldBe(HttpStatusCode.OK);
+
+        // Verify that the custom cache was used for replay detection
+        cache.SetAsyncCalls.Count.ShouldBeGreaterThan(0);
+        cache.GetAsyncCalls.Count.ShouldBeGreaterThan(0);
+    }
```

Comment:
These are the tests that are really new/of interest in this PR. They show that customization of either keyed dependency can be done.